wolfcrypt/src/ext_xmss.c

```diff
@@ -0,0 +1,970 @@
+        return BAD_FUNC_ARG;
+    }
+
+    if (key->state != WC_XMSS_STATE_PARMSET) {
+        WOLFSSL_MSG("error: XmssKey not ready for reload");
+        return -1;
+    }
+
+    if (key->write_private_key == NULL || key->read_private_key == NULL) {
+        WOLFSSL_MSG("error: XmssKey write/read callbacks are not set");
+        return -1;
+    }
+
+    if (key->context == NULL) {
+        WOLFSSL_MSG("error: XmssKey context is not set");
+        return -1;
+    }
+
+    /* Allocate sk array. */
+    ret = wc_XmssKey_AllocSk(key);
+
+    if (ret != 0) {
+        return ret;
+    }
+
+    /* Read the current secret key from NV storage. Force clear it
+     * immediately. This is just to sanity check the secret key
+     * is readable from permanent storage. */
+    cb_rc = key->read_private_key(key->sk, key->sk_len, key->context);
+    ForceZero(key->sk, key->sk_len);
+
+    if (cb_rc != WC_XMSS_RC_READ_TO_MEMORY) {
+        WOLFSSL_MSG("error: XMSS read from NV storage failed");
+        key->state = WC_XMSS_STATE_BAD;
+        return -1;
+    }
+
+    key->state = WC_XMSS_STATE_OK;
+
+    return 0;
+}
+
+/* Gets the XMSS/XMSS^MT private key length.
+ *
+ * Parameters must be set before calling this, as the key size (sk_bytes)
+ * is a function of the parameters.
+ *
+ * Note: the XMSS/XMSS^MT private key format is implementation specific,
+ * and not standardized. Interoperability of XMSS private keys should
+ * not be expected.
+ *
+ *  key         [in]      The XMSS key.
+ *  len         [out]     The length of the private key in bytes.
+ *
+ *  returns     0 on success.
+ *  returns     BAD_FUNC_ARG when a parameter is NULL.
+ *  returns     -1 on sign fail.
+ * */
+int wc_XmssKey_GetPrivLen(const XmssKey * key, word32 * len)
+{
+    if (key == NULL || len == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    if (key->state != WC_XMSS_STATE_OK && key->state != WC_XMSS_STATE_PARMSET) {
+        /* params.sk_bytes not set yet. */
+        return -1;
+    }
+
+    *len = XMSS_OID_LEN + (word32) key->params.sk_bytes;
+
+    return 0;
+}
+
+/* Signs the message using the XMSS secret key, and
+ * updates the secret key on NV storage.
+ *
+ * Both operations must succeed to be considered
+ * successful.
+ *
+ * On success:  sets key state to WC_XMSS_STATE_OK.
+ * On failure:  sets key state to WC_XMSS_STATE_BAD
+ *
+ * If no signatures are left, sets state to WC_XMSS_STATE_NOSIGS.
+ */
+static void wc_XmssKey_SignUpdate(XmssKey* key, byte * sig, word32 * sigLen,
+    const byte * msg, int msgLen)
+{
+    int                ret = -1;
+    unsigned long long len = *sigLen;
+    enum wc_XmssRc     cb_rc = WC_XMSS_RC_NONE;
+
+    /* Set the key state to bad by default. State is presumed bad
+     * unless a correct sign and update operation happen together. */
+    key->state = WC_XMSS_STATE_BAD;
+    *sigLen = 0;
+
+    /* Read the current secret key from NV storage.*/
+    cb_rc = key->read_private_key(key->sk, key->sk_len, key->context);
```

Comment:
sanity check that read_private_key is not null

Reply:
Good catch! I'll add the missing checks to `wc_XmssKey_Sign()`, which is the only function that calls `static void wc_XmssKey_SignUpdate()`.